common/djangoapps/third_party_auth/provider.py

```diff
@@ -29,21 +30,43 @@ def _enabled_providers(cls):
         """
         Helper method that returns a generator used to iterate over all providers
         of the current site.
+
+        Provider configurations are partitioned on site + some key (backend
+        name in the case of OAuth, slug for SAML, and consumer key for LTI).
         """
-        oauth2_backend_names = OAuth2ProviderConfig.key_values('backend_name', flat=True)
-        for oauth2_backend_name in oauth2_backend_names:
-            provider = OAuth2ProviderConfig.current(oauth2_backend_name)
-            if provider.enabled_for_current_site and provider.backend_name in _PSA_OAUTH2_BACKENDS:
+        def current_partitioned(prov_cls, filters=None):
+            """
+            Like ConfigurationModel.current() but partitions on KEY_FIELDS and
+            allows additional filters dict. (Does not include caching.)
+            """
+            def partition_key(obj):
+                """Compute the partition key for this object."""
+                return tuple(getattr(obj, k) for k in prov_cls.KEY_FIELDS)
+
+            configs = prov_cls.objects.filter(**filters) if filters else prov_cls.objects.all()
+            # Get the most recent record in each partition
+            return unique_everseen(configs.order_by('-change_date'), partition_key)
+
+        site = Site.objects.get_current(get_current_request())
```

Comment:
It's easier to read if we have this statement before the `current_partitioned` function.

Reply:
Agreed, although my intention would be to move `current_partitioned` out to django-config-models—or more likely, to change `ConfigurationModel.current` to accept a QuerySet or a list of additional filters. It's actually not totally clear if this is necessary at all, though; I'm looking into whether we can just add `'site'` to the `KEY_FIELDS` of each model, which should fix the bug. (Might not help with the 1+N, though!)

Reply:
(Ah... I think I do want to add `'site'` to `KEY_FIELDS`, but it wouldn't be sufficient to address 1+N.)